[v0.41.0 - QA] Documentation Fixes (#7237)
PennyLaneAI/pennylane: doc/news/program_capture_sharp_bits.rst

Author: @isaacdevlugt, check it out 😎

Array([0.99875027+0.j        , 0.        +0.j        ,
       0.        +0.j        , 0.        -0.04997917j], dtype=complex64)

Gradients
---------

Currently the devices ``default.qubit`` and ``lightning.qubit`` are the only devices

> that support gradients with program capture enabled. ``default.qubit`` currently only supports

Author: ```suggestion
that support gradients with program capture enabled. ``default.qubit`` currently supports
```

``adjoint``, ``backprop`` and ``finite-diff``. ``lightning.qubit`` currently only supports 
``adjoint``. The ``parameter_shift`` method is not yet supported with program capture enabled, 
and will raise an error if used. 

.. code-block:: python

    import jax 

    qml.capture.enable() 

    dev = qml.device('default.qubit', wires=1)

    @qml.qnode(dev)
    def circuit(x):
        qml.RX(x, wires=0)
        return qml.expval(qml.Z(0))

    bp_qn = circuit.update(diff_method="backprop")

>     adj_qn = circuit.update(diff_method="adjoint")

Reviewer: @andrijapau you love the update method! 🚀

>>> x = jax.numpy.array(jax.numpy.pi / 4)
>>> jax.jacobian(bp_qn)(x)
Array(-0.70710677, dtype=float32)
>>> jax.jacobian(adj_qn)(x)
Array(-0.70710677, dtype=float32)

However, there are some limitations to be aware of 
when using ``adjoint`` with ``default.qubit``.

> Control flow and gradients
> ~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Control flow and diff_method="adjoint"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

> Control flow like ``for``, ``while`` and ``conditionals`` 

Reviewer: ```suggestion
Control flow like ``for``, ``while`` and ``if/else`` 
```

are not currently supported when using ``"adjoint"`` with ``default.qubit``.
For example, the following code will raise an error:

.. code-block:: python

    import jax

    qml.capture.enable()

    dev = qml.device("default.qubit",wires=2)

    @qml.qnode(dev, diff_method="adjoint")
    def f(x):
        for i in range(2):
            qml.RX(x, wires=i)
        return qml.expval(qml.Z(0))

>>> x = jax.numpy.array(jax.numpy.pi / 4)
>>> jax.jacobian(f)(x)
NotImplementedError: Primitive for_loop does not have a jvp rule and is not supported.

> Higher-order primitives and gradients
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
Higher-order primitives and diff_method="adjoint"
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Higher-order primitives like ``qml.ctrl`` and ``qml.adjoint`` are not currently supported
when using ``"adjoint"`` with ``default.qubit``. For example, the following code will raise an error:

.. code-block:: python

    import jax

    qml.capture.enable()

    dev = qml.device("default.qubit",wires=2)

    @qml.qnode(dev, diff_method="adjoint")
    def f(x):
        qml.ctrl(qml.RX, control=0)(x, 1)
        return qml.expval(qml.Z(0))

>>> x = jax.numpy.array(jax.numpy.pi / 4)
>>> jax.jacobian(f)(x)

> NotImplementedError: Primitive for_loop does not have a jvp rule and is not supported.

Author: ```suggestion
NotImplementedError: Primitive ctrl_prim does not have a jvp rule and is not supported.
```
